components/client/src/test/unit/java/com/hotels/styx/client/OriginRestrictionLoadBalancingStrategyTest.java

```diff
@@ -15,126 +15,128 @@
  */
 package com.hotels.styx.client;
 
-import com.hotels.styx.api.HttpClient;
-import com.hotels.styx.api.HttpRequest;
 import com.hotels.styx.api.Id;
 import com.hotels.styx.api.client.Origin;
 import com.hotels.styx.api.client.RemoteHost;
-import com.hotels.styx.api.client.loadbalancing.spi.LoadBalancingStrategy;
-import com.hotels.styx.client.netty.connectionpool.StubConnectionPool;
+import com.hotels.styx.api.client.loadbalancing.spi.LoadBalancer;
+import com.hotels.styx.api.client.loadbalancing.spi.LoadBalancingMetricSupplier;
 import org.testng.annotations.Test;
 
+import java.util.HashMap;
 import java.util.List;
-import java.util.function.Function;
+import java.util.Map;
+import java.util.Optional;
 import java.util.stream.Stream;
 
-import static com.hotels.styx.api.HttpRequest.Builder.get;
-import static com.hotels.styx.api.Id.GENERIC_APP;
 import static com.hotels.styx.api.client.Origin.newOriginBuilder;
 import static com.hotels.styx.api.client.RemoteHost.remoteHost;
-import static java.util.function.Function.identity;
 import static java.util.stream.Collectors.toList;
 import static org.hamcrest.MatcherAssert.assertThat;
-import static org.hamcrest.Matchers.contains;
-import static org.hamcrest.Matchers.emptyIterable;
+import static org.hamcrest.Matchers.greaterThan;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isOneOf;
 import static org.mockito.Mockito.mock;
 
 public class OriginRestrictionLoadBalancingStrategyTest {
     List<RemoteHost> origins = Stream.of(0, 1, 2, 3, 4, 5, 6)
-            .map(this::remoteHostFromPort)
+            .map(i -> newOriginBuilder("localhost", 8080 + i).id("origin-" + i).build())
+            .map(origin -> remoteHost(origin, mock(StyxHostHttpClient.class), mock(LoadBalancingMetricSupplier.class)))
             .collect(toList());
 
-    OriginRestrictionLoadBalancingStrategy strategy = new OriginRestrictionLoadBalancingStrategy(
-            () -> origins, (context) -> origins, "originRestrictionCookie");
+    private final LoadBalancer delegate = (preferences) -> Optional.of(origins.get(0));
+
+    OriginRestrictionLoadBalancingStrategy strategy = new OriginRestrictionLoadBalancingStrategy(() -> origins, delegate);
 
     @Test
     public void shouldDisregardRestrictionCookieValueIfNotValid() {
-        Iterable<RemoteHost> partition = strategy.vote(contextWith(request ->
-                request.addCookie("originRestrictionCookie", "*-01")));
-
-        assertThat(partition, contains(origins.toArray()));
+        for (int i = 0; i < 10; i++) {
```

Comment:
I believe the code has evolved since I fixed these tests. I will revisit this part.

Reply:
Hi Kyle. This is because the load balancing API has changed, and `OriginsRestrictionLoadBalancer` had to change along with it. Styx client now passes in a `LoadBalancer.Preferences` object to indicate that an origin restriction cookie was present. The `OriginsRestrictionLoadBalancer` now has to choose an origin according to that cookie (or LB.Preferences).

So whenever there is a tie (that is, the cookie matches with multiple origins), then the `OriginsRestrictionLoadBalancer` simply chooses one randomly. 

This is why you'll see the test repeating themselves several times. It ensures they don't pass accidentally. They need to pass consistently. For example in:

```
        for (int i = 0; i < 10; i++) {
            Optional<RemoteHost> chosenOne = strategy.choose(lbPreference(Optional.of("origin-1")));
            assertThat(chosenOne.get(), is(origins.get(1)));
        }
```

This ensures that `origin-1` is always returned. Not just by pure luck.

Hope this helps.

Reply:
As a clarification to above. The above behaviour applies only when the origins restriction cookie is present. All other times the decision is delegated to the configured load balancer.

Reply:
It would be nice to know what is special about the number 10 though. Why does 10 ensure it, but not 9 (for example)?

Reply:
There is nothing special about number 10. It is just nice round number. We could equally well run this 9 times or 11 times. I know what you are getting at. But how would you improve from this? Make 10 a constant?

Reply:
What I mean is, do we not have a way of testing it that *definitely* ensures it works? It sounds like we are just using a high number to reduce the probability that a mistake will slip through, when ideally we want the test to be certain.
If it's due to randomness, maybe we could use a constant seed?

Reply:
You are correct. I rewrote these tests with mock RNG. Have another look.